nemo/collections/speechlm/models/speech_to_text_llm_model.py

```diff
@@ -0,0 +1,1127 @@
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+
+import json
+import os
+from dataclasses import dataclass
+from types import MethodType
+from typing import Any, Callable, Dict, List, Optional, Tuple, Union
+
+import lightning.pytorch as L
+import torch
+import torch.nn as nn
+from megatron.core import dist_checkpointing, parallel_state, tensor_parallel
+from megatron.core.inference_params import InferenceParams
+from megatron.core.num_microbatches_calculator import (
+    get_current_global_batch_size,
+    get_num_microbatches,
+    reconfigure_num_microbatches_calculator,
+)
+from megatron.core.optimizer import OptimizerConfig
+from megatron.core.transformer import MegatronModule
+from megatron.core.transformer.enums import ModelType
+from megatron.core.transformer.transformer_config import TransformerConfig
+from omegaconf import DictConfig, ListConfig
+from pytorch_lightning.utilities import rank_zero_only
+
+from nemo.collections.asr.models import ASRModel
+from nemo.collections.asr.parts.utils.eval_utils import remove_punctuations
+from nemo.collections.common.data.utils import move_data_to_device
+from nemo.collections.common.metrics import MetricStringToTorchMetric, TextMetricsSet
+from nemo.collections.common.tokenizers.tokenizer_spec import TokenizerSpec
+from nemo.collections.llm import fn
+from nemo.collections.llm.gpt.model.base import (
+    GPTConfig,
+    GPTModel,
+    get_batch_on_this_context_parallel_rank,
+    get_packed_seq_params,
+)
+from nemo.collections.nlp.modules.common.megatron.utils import build_position_ids
+from nemo.collections.nlp.parts.utils_funcs import get_last_rank
+from nemo.collections.speechlm.models.base import SpeechLanguageModel
+from nemo.collections.speechlm.modules.asr_module import ASRModuleConfig
+from nemo.collections.speechlm.modules.modality_adapter import ModalityAdapterConfig
+from nemo.collections.speechlm.utils.text_generation.audio_text_generation_strategy import (
+    SpeechToTextGenerationStrategy,
+)
+from nemo.collections.speechlm.utils.text_generation.audio_text_generation_utils import (
+    clean_end_string,
+    default_inference_config,
+    generate,
+    get_computeprob_response,
+)
+from nemo.lightning import io
+from nemo.lightning.io.pl import ckpt_to_weights_subdir
+from nemo.lightning.pytorch.optim import MegatronOptimizerModule, OptimizerModule
+from nemo.utils import AppState, logging, model_utils
+
+
+def set_input_tensor(self, tensor: torch.Tensor):
+    """
+    Placeholder function for pipeline parallel, not implemented yet.
+    """
+    pass
+
+
+def speech_to_text_llm_data_step(dataloader_iter) -> Dict[str, Any]:
```

Comment:
Can you document where this is used and in what context?

Reply:
added doc. in nemo2 each model config needs to have its own data step func and forward func...